extension/src/openvic-extension/singletons/CursorSingleton.hpp

```diff
@@ -0,0 +1,99 @@
+#pragma once
+
+#include <godot_cpp/core/class_db.hpp>
+#include <godot_cpp/core/object.hpp>
+#include <godot_cpp/classes/file_access.hpp>
+#include <godot_cpp/classes/image_texture.hpp>
+#include <godot_cpp/classes/input.hpp>
+#include <godot_cpp/templates/vector.hpp>
+#include <godot_cpp/variant/string.hpp>
+#include <godot_cpp/variant/string_name.hpp>
+
+#include <openvic-simulation/dataloader/Dataloader.hpp>
+#include <openvic-extension/singletons/GameSingleton.hpp>
+#include <openvic-simulation/types/OrderedContainers.hpp>
+#include <openvic-simulation/types/IdentifierRegistry.hpp>
+
+namespace OpenVic {
+
+	class CursorSingleton : public godot::Object {
+		
+		/*
+		This singleton only handles the dataloading for the cursors.
+		The functionality for actually setting one of these as a cursor
+		and animating animated cursors is done by the CursorManager autoloaded
+		gd script. With the exception of the shapeMap and its functions??
+		*/
+
+		GDCLASS(CursorSingleton, godot::Object);
+		static inline CursorSingleton* _singleton = nullptr;
+
+		//an intermediate data type to help with loading cursors
+		//the size of images/hotspots corresponds to resolutionsPerCursor
+		struct image_hotspot_pair_asset_t {
+			std::vector<godot::Vector2i> hotspots;
+			std::vector<godot::Ref<godot::ImageTexture>> images;
+		};
+
+		//.cur files use all but the last 2 properties, rest are for .ani
+		//use vector2i instead of ref<vector2i> because of "unreference is not a member of godot::vector2i"
+		//images is an array of ImageTexture, just godot doesn't like typed arrays with image texture for some reason
+		struct cursor_asset_t {
+			std::vector<godot::TypedArray<godot::Vector2i>> hotspots;
+			std::vector<godot::Array> images;
+			godot::TypedArray<godot::Vector2i> resolutions;
+			int animationLength; //1 for static cursors
+			std::optional<godot::TypedArray<float>> displayRates;
+			std::optional<godot::TypedArray<int>> sequence;
```

Comment:
Godot has special `Packed***Array` types that are faster and more space-efficient than regular arrays full of the same value type:
- `PackedByteArray` - array of `uint8_t`s
- `PackedColorArray` - array of `Color`s (so basically `Vector4`s)
- `PackedFloat32Array` - array of `float`s
- `PackedFloat64Array` - array of `double`s
- `PackedInt32Array` - array of `int32_t`s
- `PackedInt64Array` - array of `int64_t`
- `PackedStringArray` - array of `godot::String`s
- `PackedVector2Array` - array of `Vector2`s (2 floats)
- `PackedVector3Array` - array of `Vector3`s (3 floats)
- `PackedVector4Array` - array of `Vector4`s (4 floats)

So here it would be best to use `godot::PackedFloat32Array` and `godot::PackedInt32Array`. Unfortunately there isn't a `PackedVector2iArray` we could use for `resolutions`, so that will have to stay as a regular array (unless you're able to switch to using float-based `Vector2`s? Tbh probably not worth it anyway).

Reply:
Made the change to PackedFloat and Ints. You probably could use Vector2s instead of Vector2is, so I'll leave this unresolved in case you really do want that.